Add Swift IPC layout test

adetaylor opened the pull request:
#### 18770b8cdeffcaf19ead1809dc694b1d8bd37f1a
<pre>
Add Swift IPC layout test
<a href="https://bugs.webkit.org/show_bug.cgi?id=303989">https://bugs.webkit.org/show_bug.cgi?id=303989</a>
<a href="https://rdar.apple.com/166293273">rdar://166293273</a>

Reviewed by Richard Robinson.

We previously landed an ability for CoreIPC to call directly into Swift message
handlers, and we added a test message handler within the IPCTester
infrastructure. However, we landed no automated test to exercise this new code.
That&apos;s what this PR does.

This will only have any effect in builds where ENABLE_IPC_TESTING is true, which
as standard is ASAN or debug builds.

In all such builds, we&apos;ll now exercise a code path to call into C++ and receive
an asynchronous reply. If additionally ENABLE_IPC_TESTING_SWIFT is enabled,
we&apos;ll also call into a Swift message receiver.

As well as adding the new layout test (HTML) it also makes a tiny change to the
recently-added IPCTester logic, to ensure the C++ logic behaves slightly
differently to the Swift logic, as an extra layer of proof that the Swift
receiver was called. It also adds a missing call to removeMessageReceiver.

Test: ipc/swift-receiver.html
Canonical link: <a href="https://commits.webkit.org/304693@main">https://commits.webkit.org/304693@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c1f35f12f970fa3252d4e60961a41c0b167e2b2b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136271 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8628 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47551 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143982 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/89242 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/fbc84c72-6c97-4021-bbf9-01b520472a81) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9306 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8472 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104193 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/89242 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/5b56c5ad-de64-4a8c-a9d5-61252d8be547) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139216 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6766 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122106 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85026 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/87a28b7a-63ef-448a-9073-442850076dbb) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6421 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4114 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4574 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115711 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40307 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146726 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8310 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40875 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112532 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8327 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6982 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112876 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28659 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6356 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118411 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/62297 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8358 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36468 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8076 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71917 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8298 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8150 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on LayoutTests/ipc/swift-receiver.html:
```diff
@@ -0,0 +1,25 @@
+<script src="../resources/testharnessreport.js"></script>
+<body>
+<script>
+promise_test(async (t) => {
```
not really specific to this PR but I’m just curious why these types of tests are layout tests instead of API tests given they’re 100% logic based and not really visual

adetaylor replied:
Yeah I agree... it's a bit odd isn't it.

jon-b-apple commented on LayoutTests/ipc/swift-receiver.html:
```diff
@@ -0,0 +1,41 @@
+promise_test(async t => {
+    if (!window.IPC)
+        return;
+    const testerID = 447;
```
It seems like this value isn't used?

adetaylor replied:
Good spot, copypasta, removed!

jon-b-apple commented on Source/WebKit/Shared/IPCTester.cpp:
```diff
@@ -191,8 +191,10 @@ void IPCTester::sendAsyncMessageToSwiftReceiver(IPC::Connection& connection, uin
 #if ENABLE(IPC_TESTING_SWIFT)
     connection.sendWithAsyncReply(Messages::IPCTesterReceiverSwift::AsyncMessage(arg0 + 1), [arg0](uint32_t newArg0) {
-        ASSERT_UNUSED(arg0, newArg0 == arg0 + 2);
+        ASSERT_UNUSED(arg0, newArg0 == arg0 + 3);
```
When we're testing in Release+ASAN, would this ASSERT still fire? I'm wondering if we should be checking the value we get back in the test anyway, even if we validate with asserts in the reply handler.

adetaylor replied:
Hmmmm.
Yeah I think I've made too many short cuts in trying to re-use the existing `AsyncMessage` infrastructure, and you're probably right, I should add something more directly relevant to the needs of testing Swift here, even if it's a bit more code. I'll see what I can do.

adetaylor replied:
OK I had a crack at rearranging things so that the return value is tested in the actual JS test - please take a look.

jon-b-apple commented on LayoutTests/ipc/swift-receiver.html:
```diff
@@ -0,0 +1,35 @@
+        const defaultTimeout = 1000;
+
+        let result;
+        let connection = IPC.connectionForProcessTarget(processTarget);
```
I think this could still be simplified a little since we're always sending to the UI process, maybe `IPC.sendMessage('UI', 0, <message name + params>` would work here?

adetaylor replied:
OK, that didn't quite work, because it provided no means to wait for the reply, but thanks for the poke - I found a routine `IPC.IPC.sendWithPromisedReply` in an existing helper `ipc.js` so I'm now using that and the test is quite a bit simpler.

jon-b-apple reviewed (commented):
Looks good to me, with some possible simplifications in the testcase structure if you'd like

adetaylor committed:
Add Swift IPC layout test
https://bugs.webkit.org/show_bug.cgi?id=303989
rdar://166293273

Reviewed by Richard Robinson.

We previously landed an ability for CoreIPC to call directly into Swift message
handlers, and we added a test message handler within the IPCTester
infrastructure. However, we landed no automated test to exercise this new code.
That's what this PR does.

This will only have any effect in builds where ENABLE_IPC_TESTING is true, which
as standard is ASAN or debug builds.

In all such builds, we'll now exercise a code path to call into C++ and receive
an asynchronous reply. If additionally ENABLE_IPC_TESTING_SWIFT is enabled,
we'll also call into a Swift message receiver.

As well as adding the new layout test (HTML) it also makes a tiny change to the
recently-added IPCTester logic, to ensure the C++ logic behaves slightly
differently to the Swift logic, as an extra layer of proof that the Swift
receiver was called. It also adds a missing call to removeMessageReceiver.

Test: ipc/swift-receiver.html
Canonical link: https://commits.webkit.org/304693@main
